[RISCV] Tune flag for fast vrgather.vv

ppenzin opened the pull request:
Add tune knob for N*Log2(N) vrgather.vv cost.

ppenzin committed:
[RISCV] Tune flag for fast vrgather.vv

Add tune knob for N*Log2(N) vrgather.vv cost.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2848,6 +2848,11 @@ InstructionCost RISCVTargetLowering::getLMULCost(MVT VT) const {
 /// operand (index and possibly mask) are handled separately.
 InstructionCost RISCVTargetLowering::getVRGatherVVCost(MVT VT) const {
+  auto LMULCost = getLMULCost(VT);
+  if (true && Subtarget.hasFastVRGather() && LMULCost.isValid()) {
```
Delete `true && `

ppenzin replied:
Addressed, I guess I should've checked the change _after_ applying it to my public fork...

lukel97 reviewed (commented):
Thanks, could you add a new RUN line in test/Analysis/CostModel/RISCV/rvv/shuffle-permute that has the fast flag enabled?

ppenzin committed:
Fix typo

ppenzin committed:
Add run line to RISCV shuffle-permute.ll test

ppenzin commented:
@lukel97 does this work?

ppenzin commented:
I took a brief look, it doesn't look like it would be very difficult to convert `undef` into arguments, I can give that a go in a separate PR.

lukel97 commented on llvm/test/Analysis/CostModel/RISCV/shuffle-permute.ll:
```diff
@@ -3,6 +3,8 @@
 ; RUN: opt < %s -passes="print<cost-model>" 2>&1 -disable-output -mtriple=riscv32 -mattr=+v,+f,+d,+zfh,+zvfhmin | FileCheck %s
 ; RUN: opt < %s -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output -mtriple=riscv32 -mattr=+v,+f,+d,+zfh,+zvfh | FileCheck %s --check-prefix=SIZE
 ; RUN: opt < %s -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output -mtriple=riscv32 -mattr=+v,+f,+d,+zfh,+zvfhmin | FileCheck %s --check-prefix=SIZE
+; RUN: opt < %s -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output -mtriple=riscv32 -mattr=+v,+f,+d,+zfh,+zvfh,+fast-vrgather | FileCheck %s --check-prefix=FAST-VRG
```
Can you drop the cost-kind=code-size argument so it tests the throughput cost instead?

ppenzin replied:
Addressed

lukel97 commented on llvm/lib/Target/RISCV/RISCVProcessors.td:
```diff
@@ -490,6 +490,7 @@ def TENSTORRENT_ASCALON_D8 : RISCVProcessorModel<"tt-ascalon-d8",
                                                   FeatureUnalignedScalarMem,
                                                   FeatureUnalignedVectorMem]),
                                                  [TuneNoDefaultUnroll,
+                                                  TuneFastVRGather,
```
Looks like an extraneous new line

ppenzin replied:
Joining with next line would definitely push that past 80 columns, is that OK?

lukel97 replied:
Ah ignore my comment, my browser window had truncated the rest of the line, sorry!

lukel97 reviewed (commented):
I think you can ignore the undef warning for now, pretty much every test in that directory is guilty of using undef!

ppenzin committed:
Remove code size from analysis test

lukel97 reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2848,6 +2848,11 @@ InstructionCost RISCVTargetLowering::getLMULCost(MVT VT) const {
+    unsigned Log = Log2_64(*LMULCost.getValue());
+    return LMULCost * Log;
+  }
   return getLMULCost(VT) * getLMULCost(VT);
```
Can we reuse the `LMULCost` local here? No reason to call it 3 times.

ppenzin replied:
Good point, addressed

ppenzin committed:
Reuse calculated LMUL cost

preames reviewed (approved):
LGTM

ppenzin commented:
Since there were a few approvals, I've removed WIP. @topperc what do you think?

topperc commented on llvm/test/Analysis/CostModel/RISCV/shuffle-permute.ll:
```diff
@@ -44,6 +46,24 @@ define void @general_permute_single_source() {
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %v4f32 = shufflevector <4 x float> undef, <4 x float> undef, <4 x i32> <i32 3, i32 2, i32 1, i32 1>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 6 for instruction: %v8f32 = shufflevector <8 x float> undef, <8 x float> undef, <8 x i32> <i32 7, i32 6, i32 5, i32 5, i32 3, i32 2, i32 1, i32 0>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 5 for instruction: %v4f64 = shufflevector <4 x double> undef, <4 x double> undef, <4 x i32> <i32 3, i32 2, i32 3, i32 0>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 0 for instruction: ret void
```
Should we test larger LMUL? This is only testing up to LMUL=2?

ppenzin replied:
We are testing `16 x i64` and `16 x double`, which gets LMULCost of 8; with quadratic estimate it leads to total cost of 139 and with the fast flag it is 59.

topperc commented on llvm/test/Analysis/CostModel/RISCV/shuffle-permute.ll:
```diff
@@ -44,6 +46,24 @@ define void @general_permute_single_source() {
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 6 for instruction: %v16i16 = shufflevector <16 x i16> undef, <16 x i16> undef, <16 x i32> <i32 15, i32 14, i32 13, i32 11, i32 11, i32 11, i32 9, i32 8, i32 7, i32 6, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 3 for instruction: %v4i32 = shufflevector <4 x i32> undef, <4 x i32> undef, <4 x i32> <i32 3, i32 2, i32 2, i32 0>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 6 for instruction: %v8i32 = shufflevector <8 x i32> undef, <8 x i32> undef, <8 x i32> <i32 7, i32 4, i32 5, i32 4, i32 3, i32 2, i32 1, i32 0>
+; FAST-VRG-NEXT:  Cost Model: Found an estimated cost of 5 for instruction: %v4i64 = shufflevector <4 x i64> undef, <4 x i64> undef, <4 x i32> <i32 3, i32 1, i32 1, i32 0>
```
Why is 256-bit i64 vector cheaper than 256-bit i32 and i16 vectors?

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVFeatures.td:
```diff
@@ -1365,6 +1365,10 @@ def FeatureUnalignedVectorMem
 
+def TuneFastVRGather
+   : SubtargetFeature<"fast-vrgather", "HasFastVRGather",
+                      "true", "Has vrgather.vv with LMUL*log2(LMUL) latency">;
```
Not blocking, but what if the `vrgather.vv` is not with `LMUL*log2(LMUL)` complexity?

ppenzin replied:
IDK, that is the only other alternative I know of personally. We can rename flag to be something like 'HasLog2Vrgather' instead of just "fast".

wangpc-pp replied:
What about adding an enum to represent the cost model?
```cpp
enum VRGatherCostModel {
    Quadratic, // Default
    Log2,
    ....
}
```
```
def TuneLog2VRGather
   : SubtargetFeature<"log2-vrgather", "VRGatherCostModel",
                      "Log2", "Has vrgather.vv with LMUL*log2(LMUL) latency">;
```
This is more extensible I think. I don't know the details but I remember that the Ventana's V2 has an optimized vrgather.vv implementation as well.

ppenzin replied:
I think I can do that. @topperc and @lukel97 any thoughts from maintainability point of view? Also, anyone from Ventana would like to share their input? (@mgudim ?)

lukel97 replied:
How many different types of cost model would we have? I would have thought `LMUL * log2(LMUL)` would be as good as a uarch can get. I'm not strongly opinionated on this though, an enum sounds reasonable to me.

preames replied:
I don't have a strong preference here, and would tend to bias towards support what we know of now, then change later if multiple options show up.

topperc replied:
x280 cost is linear in VL. Later x280 is constant time for VL<= (VLEN/2)/SEW (single DLEN), but linear in VL for everything else. Later x280 is quadratic in LMUL*2 except for fractional LMUL or when element at a time would be faster (i.e. there are less elements in a destination DLEN than the number of possible source DLENs for them).

ppenzin replied:
Enum it is then, I suppose :smile:

ppenzin replied:
That should be addressed, PTAL

arcbbb commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2848,7 +2848,12 @@ InstructionCost RISCVTargetLowering::getLMULCost(MVT VT) const {
-  return getLMULCost(VT) * getLMULCost(VT);
+  auto LMULCost = getLMULCost(VT);
+  if (Subtarget.hasFastVRGather() && LMULCost.isValid()) {
+    unsigned Log = Log2_64(*LMULCost.getValue());
+    return LMULCost * Log;
```
Would the cost be zero if LMULCost is 1?

ppenzin replied:
Yes, log will return zero if cost is 1, and will return -1 for zero, though `vrgather` is unlikely to have a cost of 1. I can guard this to only produce the number if log value is positive, what do you think?

ppenzin replied:
Addressed

arcbbb replied:
LGTM

ppenzin committed:
Guard against zero and negative log values

ppenzin committed:
Change vrgather cost model to use an enum

Rename option to explicitly say "log", change tune flag to set one of
the values.

ppenzin committed:
Update comment

ppenzin commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2848,7 +2848,13 @@ InstructionCost RISCVTargetLowering::getLMULCost(MVT VT) const {
 /// is generally quadratic in the number of vreg implied by LMUL.  Note that
```
Addressed

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -2845,10 +2845,19 @@ InstructionCost RISCVTargetLowering::getLMULCost(MVT VT) const {
 /// Return the cost of a vrgather.vv instruction for the type VT.  vrgather.vv
-/// is generally quadratic in the number of vreg implied by LMUL.  Note that
+/// may be quadratic in the number of vreg implied by LMUL, and is assumed to
+/// be by default.  VRGartherCostModel reflects available options.  Note that
```
VRGarther -> VRGather

ppenzin replied:
Addressed.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -4228,7 +4237,7 @@ static SDValue lowerBUILD_VECTOR(SDValue Op, SelectionDAG &DAG,
   // The default lowering is to use the stack.  The vector store + scalar loads
   // is linear in VL.  However, at high lmuls vslide1down and vslidedown end up
-  // being (at least) linear in LMUL.  As a result, using the vslidedown
+  // being (at least) linear in LMUL.  As a resultdedown
```
Something bad happened to this comment.

ppenzin replied:
Indeed, good catch. Should be addressed.

ppenzin committed:
Address typos

wangpc-pp reviewed (approved):
LGTM.

topperc reviewed (approved):
LGTM

mshockwave commented:
> Can anyone help me merge this? I can rebase if needed.

Could you fix the test failure on test/CodeGen/RISCV/features-info.ll? I think that is relevant

ppenzin committed:
Fix test flag spelling, add to release notes

ppenzin commented:
Should be addressed now, but I did add it a line to release notes which requires a rebase on top of current main.

mshockwave commented:
> Should be addressed now, but I did add it a line to release notes which requires a rebase on top of current main.

LGTM, please rebase then I'll help you to merge

mshockwave commented:
> Can anyone help me merge this? I can rebase if needed.

I think there are several patches under your belt now. Feel free to request for commit access.
